blocks/autocompleters/index.js

```diff
@@ -51,55 +63,54 @@ import BlockIcon from '../block-icon';
  * @typedef {Object} Completer
  * @property {?String} className A class to apply to the popup menu.
  * @property {String} triggerPrefix the prefix that will display the menu.
- * @property {FnGetOptions} getOptions get the block options in a resolved promise.
+ * @property {FnGetOptions} getOptions get the completer options in a resolved promise.
+ * @property {?FnGetOptionKeywords} getOptionKeywords get the keywords for a given option.
+ * @property {FnRenderOptionLabel} renderOptionLabel render the label for a given option.
  * @property {?FnAllowNode} allowNode filter the allowed text nodes in the autocomplete.
  * @property {?FnAllowContext} allowContext filter the context under which the autocomplete activates.
- * @property {FnOnSelect} onSelect
+ * @property {FnGetOptionResult} getOptionResult
  */
 
 /**
  * Returns an "completer" definition for selecting from available blocks to replace the current one.
  * The definition can be understood by the Autocomplete component.
  *
- * @param {Function} onReplace  Callback to replace the current block.
- *
  * @return {Completer} Completer object used by the Autocomplete component.
  */
-export function blockAutocompleter( { onReplace } ) {
-	// Prioritize common category in block type options
-	const options = sortBy(
-		getBlockTypes(),
-		( { category } ) => 'common' !== category
-	).map( ( blockType ) => {
-		const { name, title, icon, keywords = [] } = blockType;
-		return {
-			value: name,
-			label: [
-				<BlockIcon key="icon" icon={ icon } />,
-				title,
-			],
-			keywords: [ ...keywords, title ],
-		};
-	} );
-
-	const getOptions = () => Promise.resolve( options );
+export function blockAutocompleter() {
+	let promisedOptions;
+
+	const getOptions = () => promisedOptions || ( promisedOptions = Promise.resolve(
+		// Prioritize common category in block type options
+		sortBy(
+			getBlockTypes(),
+			( { category } ) => 'common' !== category
+		)
+	) );
+	const getOptionKeywords = ( { title, keywords = [] } ) => [ ...keywords, title ];
+	const renderOptionLabel = ( { icon, title } ) => [
+		<BlockIcon key="icon" icon={ icon } />,
+		title,
+	];
 
 	const allowContext = ( before, after ) => {
 		return ! ( /\S/.test( before.toString() ) || /\S/.test( after.toString() ) );
 	};
 
-	const onSelect = ( blockName ) => {
-		onReplace( createBlock( blockName ) );
-	};
+	const getOptionResult = blockData => createBlock( blockData.name );
```

Comment:
I don't know this code so just wanted to double check why to remove `onReplace` here.

Reply:
It started with a dependency issue when moving completers to block registration. We didn't yet have an `onReplace`.

Then I thought completers should be more declarative anyway. The presence of an `onSomething` name made me think about that. Why should a completer know that it is replacing something in the editor? The idea is that the editor has a better idea of what it wants to do with the result of a completer selection.

Reply:
I think it ensures we remove the existing paragraph block.

Reply:
The `Autocomplete` component now calls `onReplace` instead when a block is returned.
https://github.com/WordPress/gutenberg/pull/4609/commits/ae62091c3d822703c18fd21811dcc06696023f8a#diff-8d7b3f0b32214bb1340578e3be07ed96R176

The idea is that a completer should not be involved in making actual changes to the content.